text/2020-09-03-factored-prefix.md

```diff
@@ -0,0 +1,71 @@
+# Factored Key Prefix
+
+## Summary
+
+TiKV provides a facility where the client can set a key prefix that will be used for all keys in its requests.
+
+## Motivation
+
+TiKV users need to use key prefixes to avoid conflicts.
+In keeping with DRY we should not require a program to add the same prefix everywhere.
+This will also reduce the payload sent to TiKV.
+
+## Detailed design
+
+### Prefix Manipulation
+
+TiKV already adds a first byte prefix to all keys.
+TiKV adds a first byte prefix (currently 'z') to all data keys. Other prefixes are used for metadata.
+
+When the factored prefix is used, TiKV prepends a first byte prefix to the factored prefix sent by the client and prepends that to all keys that are sent.
+All keys are returned with the first byte prefix and factored prefix stripped.
+
+A factored prefix of 0x00 is not allowed. This helps avoid accidentally using null as a factored prefix.
+
+### API
+
+The factored prefix applies to all keys in a connection.
+So it is set as a GRPC header "TIKV-PREFIX" with the value set to the key prefix.
+
+### Backwards compatibility
+
+Currently applications do not formally claim a prefix and there is no way to know what prefixes they might use.
+We can refer to these applications as V1 (version 1) applications. We can refer to applications that use the prefix as V2 (version 2).
+We must avoid mixing the usage of V1 and V2 applications.
+
+V1 applications store their data under a 'z' first byte prefix
```

Comment:
how about keeping the 'z' and append the user prefix after the 'z'? that way we don't have a V2.

Reply:
This can create conflicts with an existing TiDB application. We would like for all users to be able to upgrade to using key prefixes.

Reply:
It doesn't seems to me upgrade ability is required for multi-tenancy. We can always start a new TiKV instance for multi-tenancy setup. Also if for a use case user already have prefixes in the data, he can seamlessly change to have key prefixes support described in this RFC, also without the need to have the headache to have a V2.

Reply:
I have started to do some implementation on this and there are some complications.
If we can make future TiDB prefixes configurable then the situation will be safer. That is, TiDB currently uses 'm' and 't', so a user can reserve these for TiDB when upgrading. If TiDB starts using a new prefix in a new release (lets say it is 'd'), that prefix could conflict with an existing application that is added in the mean time. However, if we make the prefix configurable then the user can configure that prefix so that it doesn't conflict when they upgrade.

Reply:
So I think we can stick with using just the 'z' first byte prefix at the cost of some upgrade hassle for existing users.

Reply:
Okay, sticking with the 'z' first byte prefix. The implementation is much simpler and I think this will work out okay for users.